diff --git a/package/kernel/linux/modules/fs.mk b/package/kernel/linux/modules/fs.mk
index e3539cdde0..e7b2cd9ba4 100644
--- a/package/kernel/linux/modules/fs.mk
+++ b/package/kernel/linux/modules/fs.mk
@@ -427,7 +427,7 @@ define KernelPackage/fs-nfs-common
        $(LINUX_DIR)/fs/lockd/lockd.ko \
        $(LINUX_DIR)/net/sunrpc/sunrpc.ko \
        $(LINUX_DIR)/fs/nfs_common/grace.ko \
-       $(LINUX_DIR)/fs/nfs_common/nfs_ssc.ko
+       #$(LINUX_DIR)/fs/nfs_common/nfs_ssc.ko
   AUTOLOAD:=$(call AutoLoad,30,grace sunrpc lockd)
 endef
 
