--- a/feeds/packages/utils/irqbalance/files/irqbalance.init
+++ b/feeds/packages/utils/irqbalance/files/irqbalance.init
@@ -37,7 +37,7 @@
 	if [ -n "$banned_cpulist" ]; then
 		procd_set_param env IRQBALANCE_BANNED_CPULIST="$banned_cpulist"
 	fi
-	procd_set_param command /usr/sbin/irqbalance -f -c "$deepestcache" -t "$interval" "$banirq"
+	procd_set_param command /usr/sbin/irqbalance -f -j -c "$deepestcache" -t "$interval" "$banirq"
 	procd_set_param respawn
 	procd_close_instance
 }
