--- a/src/httpc.c	2019-03-20 17:05:32.920748553 +0800
+++ b/src/httpc.c	2019-03-20 17:13:24.897736654 +0800
@@ -1290,7 +1290,12 @@ http_client_reconnect
   if (strcasecmp(scheme, "https") == 0 || strcasecmp(scheme, "rtsps") == 0) {
     ssl = calloc(1, sizeof(*ssl));
     hc->hc_ssl = ssl;
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
     ssl->ctx   = SSL_CTX_new(SSLv23_client_method());
+#else
+    ssl->ctx   = SSL_CTX_new(TLS_client_method());
+#endif
+ 
     if (ssl->ctx == NULL) {
       tvhlog(LOG_ERR, "httpc", "%04X: Unable to get SSL_CTX", shortid(hc));
       goto err1;
